### ISSUE TITLE
Notebook has a root logging handler installed by default

### ISSUE THREAD

**Issue by aburgm** [ipython/ipython#8282]
This is a follow up on my questions in the gitter chat earlier today.

The observation is the following: In normal python, or the IPython console, there is no handler to the root logger installed until the root logger itself is used to issue a log message:

```
In [1]: import logging

In [2]: logging.getLogger().handlers
Out[2]: []

In [3]: logging.warn('Something happened!')
WARNING:root:Something happened!

In [4]: logging.getLogger().handlers
Out[4]: [<logging.StreamHandler at 0x42acef0>]
```

However, in the IPython notebook, there is the default stderr root logger installed right away:

```
In [1]: import logging

In [2]: logging.getLogger().handlers
Out[2]: [<logging.StreamHandler at 0x35eedd8>]
```

Was this choice deliberately made, or is this by accident?

Maybe I'm missing something, but I think that in the notebook, there should no handler be installed automatically, for a number of reasons:

 * It would make the default logging configuration consistent between standard python, the IPython console and the IPython notebook.
 * As soon as the user writes a log message with the root logger, the handler gets installed automatically, and therefore log messages are not easily missed.
 * With the current behaviour, a library that configures a child logger, and a handler for that child logger, might easily spam the notebook with debug messages that are supposed to only go in a log file (or elsewhere). For example, [astropy](https://github.com/astropy/astropy/pull/3198) seems to have such an issue, and I'm running into the same issue with my own library. The problem is that for such a library, there is no "clean" way around this. The library could remove the handler of the root logger when it is imported, which is hack-y. It could set its own logger's `propagate` attribute to `False`, so that the log messages are not propagated to the root logger, but this not only disables debug output from going into the notebook, but also more severe messages. Also, it prevents users from actually capturing all log output if they want to.

An alternative might be to add a configuration option that specifies a log level for the automatically added stream handler, so that it becomes possible to ignore less severe messages from showing up in the notebook automatically. But this would still make the behaviour different between IPython console and IPython notebook.

The only downside I see with making sure there is no default handler set is that some libraries/notebooks in use might rely on this behaviour and actively work around it, for example by disabling their own handlers if they detect that they are running in the ipython notebook. Such cases would likely break with such a change.

**Comment by carlodri** [comment]
I stumbled in the same question after realizing that the logging system for my library was silenced when called inside an IPython notebook, whereas it was working properly inside a normal IPython session. I'm absolutely not a Python expert, but I humbly agree with @aburgm's suggestions!

**Comment by davekirby** [comment]
I have been bitten by this too - calling a function that generates a lot of logging message results in them being written to the cell output in the browser, effectively killing it - even if the function being called is explicitly logging to a file.

**Comment by christang** [comment]
Another +1 to suggestion. I'm required to remove the default logging handler before ipython notebook recognizes the configurations for logging to a file.

**Comment by remram44** [comment]
Related to #6746?

**Comment by awan1** [comment]
+1: this gotcha pained me for a while. I figured out a [workaround](http://stackoverflow.com/a/35327111/2452770) (which is obvious in hindsight) that involves getting the logging handler and setting its level first. But bringing notebook in line with IPython and with general intuition (if I haven't set a handler up there shouldn't be one) seems like a good idea.

**Comment by rsdenijs** [comment]
+ 1 Stumbled on this one too. 
- 1000 for surprise behaviour!

**Comment by whoopnip** [comment]
+1, this was messing me up understanding how logging works until I realized this issue.

This is my fix:

    #Delete Jupyter notebook root logger handler
    logger = logging.getLogger()
    logger.handlers = []

    #Create logger as usual
    logger = logging.getLogger(__name__)

**Comment by takluyver** [comment]
We just released ipykernel 4.4 today, which should fix this issue. If you upgrade ipykernel with pip (a conda package should appear in a few days), you shouldn't need any workarounds.

I'll close this issue, we can reopen it if there's still something to fix with ipykernel >= 4.4.
